File: src/main/java/inno/jago/ast/SignatureNode.java
package inno.jago.ast;

import inno.jago.ast.entity.Entity;
import inno.jago.ast.parameter.ParametersNode;
import lombok.Getter;
import lombok.Setter;

/*
 * signature           :   parameters ( result )?;
*/
@Getter
@Setter
public class SignatureNode extends Entity {
    private ParametersNode parameters;
    private ResultNode result;

    public SignatureNode(ParametersNode parameters) {
        if (parameters == null) {
            System.out.println("parameters should not be null");
            return;
        }
        this.parameters = parameters;
        this.parameters.setParent(this);
        Entity.reportParsing("SIGNATURE");
    }

    public SignatureNode(ParametersNode parameters, ResultNode result) {
        this(parameters);
        if (result != null) {
            result.setParent(this);
        }
    }
}

File: src/main/kotlin/inno/jago/ast/signature/SignatureNode.kt
package inno.jago.ast.signature

import inno.jago.ast.type.TypeNode

sealed class SignatureNode<T>(
    val parameterNodes: List<ParameterNode>,
    val type: T
) {
    class TypedResultSignatureNode(
        parameterNodes: List<ParameterNode>,
        type: TypeNode
    ) : SignatureNode<TypeNode>(
        parameterNodes,
        type
    )

    class ParametrizedResultSignature(
        parameterNodes: List<ParameterNode>,
        type: List<ParameterNode>
    ) : SignatureNode<List<ParameterNode>>(
        parameterNodes,
        type
    )
}
